server/src/services/calculateAchievementsStatus.ts

```diff
@@ -0,0 +1,35 @@
+import { CalculateAchievementsStatus } from '../types/interfaces';
+
+export const calculateAchievementsStatus: CalculateAchievementsStatus = (
+  challengeId,
+  challenges,
+  achievements,
+  tasksStatus
+) => {
+  const calculatedStatuses = new Map();
+  const passedTasks = [];
+
+  for (const key in tasksStatus) {
+    if (tasksStatus[key].state === 'Success') {
+      passedTasks.push(tasksStatus[key]);
+    }
+  }
+
+  const currentChallenge = challenges.find(
+    (i) => i.challengeId === challengeId
+  );
+
+  if (!currentChallenge) {
+    throw new Error('required challenge does not exsit!');
+  }
+
+  const actualAchievementsList = currentChallenge.achievementsStatus;
```

Comment:
Why would you suggest to make an actual achievement part of the challenge and not include the `checkComplete` function into it? Wouldn't it be much more convenient this way?